ultravox/tools/gradio_demo.py

```diff
@@ -37,116 +38,122 @@ def main():
+        if "<|audio|>" not in prompt:
+            prompt += "<|audio|>"
+        sample = datasets.VoiceSample.from_prompt_and_file(prompt, audio)
+    else:
+        # Note that prompt will be "" here, since we cleared it in add_text.
+        # Instead, we can just get it from the chat history.
+        sample = datasets.VoiceSample.from_prompt(chatbot[-1][0])
+
+    if len(sample.messages) != 1:
+        raise ValueError(
+            f"Expected exactly 1 message in sample but got {len(sample.messages)}"
         )
-        reset.click(gradio_reset, [], [chatbot, prompt, audio], queue=False)
-        demo.load(gradio_reset, [], [chatbot, prompt, audio], queue=False)
 
-    demo.launch(share=True)
+    output = inference.infer_stream(
+        sample,
+        max_tokens=max_new_tokens,
+        temperature=temperature,
+    )
+    chatbot += [[None, ""]]
+    for chunk in output:
+        if isinstance(chunk, infer_base.InferenceChunk):
+            chatbot[-1][1] += chunk.text
+            yield chatbot
+
+
+def process_text(chatbot, prompt, max_new_tokens, temperature):
+    yield from process_turn(
+        chatbot, prompt, max_new_tokens=max_new_tokens, temperature=temperature
+    )
+
+
+def process_audio(chatbot, prompt, audio, max_new_tokens, temperature):
+    yield from process_turn(
+        chatbot,
+        prompt,
+        audio=audio,
+        max_new_tokens=max_new_tokens,
+        temperature=temperature,
+    )
+
+
+def gradio_reset():
+    inference.update_conversation()
+    return [], "", None
+
+
+with gr.Blocks() as demo:
+    chatbot = gr.Chatbot(scale=10, height=1000)
+
+    with gr.Row():
+        with gr.Column(scale=1):
+            reset = gr.Button("Reset")
+            audio = gr.Audio(
+                label="🎤",
+                sources=["microphone"],
+                type="filepath",
+                visible=True,
+            )
+        with gr.Column(scale=8):
+            prompt = gr.Textbox(
+                show_label=False,
+                lines=5,
+                placeholder=demo_instruction,
+                value=args.default_prompt,
+                container=True,
+            )
+        with gr.Column(scale=1):
+            max_new_tokens = gr.Slider(
+                minimum=50,
+                maximum=2000,
+                value=args.max_new_tokens,
+                step=10,
+                interactive=True,
+                label="max_new_tokens",
+            )
+            temperature = gr.Slider(
+                minimum=0,
+                maximum=5.0,
+                value=args.temperature,
+                step=0.1,
+                interactive=True,
+                label="temperature",
+            )
+    prompt.submit(add_text, [chatbot, prompt], [chatbot, prompt], queue=False).then(
+        process_text,
+        [chatbot, prompt, max_new_tokens, temperature],
+        [chatbot],
+    )
+    audio.stop_recording(add_audio, [chatbot, audio], [chatbot], queue=False).then(
```

Comment:
Also include prompt as it needs to be added to the chatbot if not empty.

```suggestion
    audio.stop_recording(add_audio, [chatbot, audio, prompt], [chatbot], queue=False).then(
```